From: Jiri Bohac <jbohac@suse.cz>
Subject: IPV6 - fix DAD race
References: bnc#577967, bnc#569916


There is a race condition in IPv6 address removal in the DAD code.
This patch is a simple workaround, which has not been agreed on upstream.
I am not 100% sure won't cause the dst entry to leak in rare cases. 

I think the dst_free is not needed, but did not get this ACKed by anyone else,
they might be going for a different fix: 
	http://kerneltrap.org/mailarchive/linux-netdev/2010/4/20/6275107

Let's keep this patch until we get a final solution upstream.

--- a/net/ipv6/addrconf.c	2010-04-17 00:12:32.000000000 +0200
+++ b/net/ipv6/addrconf.c	2010-04-20 19:07:35.000000000 +0200
@@ -3974,8 +3974,7 @@ static void __ipv6_ifa_notify(int event,
 			addrconf_leave_anycast(ifp);
 		addrconf_leave_solict(ifp->idev, &ifp->addr);
 		dst_hold(&ifp->rt->u.dst);
-		if (ip6_del_rt(ifp->rt))
-			dst_free(&ifp->rt->u.dst);
+		ip6_del_rt(ifp->rt);
 		break;
 	}
 }
